packages/dd-trace/src/appsec/api_security_sampler.js

```diff
@@ -1,61 +1,97 @@
 'use strict'
 
+const crypto = require('node:crypto')
+const LRUCache = require('lru-cache')
+const PrioritySampler = require('../priority_sampler')
+const web = require('../plugins/util/web')
 const log = require('../log')
+const { USER_KEEP, AUTO_KEEP, AUTO_REJECT, USER_REJECT } = require('../../../../ext/priority')
 
-let enabled
-let requestSampling
+const MAX_SIZE = 4096
+const DEFAULT_DELAY = 30 // 30s
 
-const sampledRequests = new WeakSet()
+let enabled
+let sampledRequests
+let prioritySampler
 
 function configure ({ apiSecurity }) {
   enabled = apiSecurity.enabled
-  setRequestSampling(apiSecurity.requestSampling)
+  const ttl = parseSampleDelay(apiSecurity.sampleDelay) * 1000
+  sampledRequests = new LRUCache({ max: MAX_SIZE, ttl })
+  prioritySampler = new PrioritySampler()
 }
 
 function disable () {
   enabled = false
+  sampledRequests?.clear()
 }
 
-function setRequestSampling (sampling) {
-  requestSampling = parseRequestSampling(sampling)
-}
+function sampleRequest (req, res) {
+  if (!enabled) return false
 
-function parseRequestSampling (requestSampling) {
-  let parsed = parseFloat(requestSampling)
+  const rootSpan = web.root(req)
+  if (!rootSpan) return false
 
-  if (isNaN(parsed)) {
-    log.warn(`Incorrect API Security request sampling value: ${requestSampling}`)
+  const priority = getSpanPriority(rootSpan)
 
-    parsed = 0
-  } else {
-    parsed = Math.min(1, Math.max(0, parsed))
+  if (priority === AUTO_REJECT || priority === USER_REJECT) {
+    return false
   }
 
-  return parsed
-}
+  if (priority === AUTO_KEEP || priority === USER_KEEP) {
+    return sample(req, res)
+  }
+
+  const isSampled = prioritySampler.isSampled(rootSpan)
 
-function sampleRequest (req) {
-  if (!enabled || !requestSampling) {
+  if (!isSampled) {
     return false
   }
 
-  const shouldSample = Math.random() <= requestSampling
+  return sample(req, res)
+}
 
-  if (shouldSample) {
-    sampledRequests.add(req)
-  }
+function sample (req, res) {
+  const key = computeKey(req, res)
+  const alreadySampled = sampledRequests.has(key)
 
-  return shouldSample
+  if (alreadySampled) return false
+
+  sampledRequests.set(key)
+
+  return true
+}
+
+function isSampled (req, res) {
+  const key = computeKey(req, res)
+  return !!sampledRequests.has(key)
+}
+
+function computeKey (req, res) {
+  const route = req.route?.path || req.url
+  const method = req.method.toLowerCase()
+  const statusCode = res.statusCode
+  const str = route + statusCode + method
+  return crypto.createHash('md5').update(str).digest('hex')
+}
+
+function parseSampleDelay (delay) {
```

Comment:
Usually config property values are resolved in the `config.js` file.
There you can set the default value, and the value read from env vars, etc

Reply:
Got it, I think there is no need for this function then